File: scripts/normalize_blog_asset_extensions.R
args <- commandArgs(trailingOnly = TRUE)

root_paths <- if (length(args) == 0) c("index.qmd", "posts") else args

all_files <- unlist(lapply(root_paths, function(path) {
  if (dir.exists(path)) {
    list.files(path, pattern = "\\.qmd$", recursive = TRUE, full.names = TRUE)
  } else if (file.exists(path)) {
    path
  } else {
    character()
  }
}), use.names = FALSE)

all_files <- unique(all_files)

pattern <- paste0(
  "(https://storage\\.googleapis\\.com/blogs_josa/sport/blog_assets/",
  "[^[:space:]\\\"')\\]>}]+/img/thumbnail/",
  "[^[:space:]\\\"')\\]>}]+)\\.(png|jpeg)"
)

for (file in all_files) {
  lines <- readLines(file, warn = FALSE, encoding = "UTF-8")
  updated <- gsub(pattern, "\\1.jpg", lines, perl = TRUE)

  if (!identical(lines, updated)) {
    writeLines(updated, file, useBytes = TRUE)
    cat("Updated", file, "\n")
  }
}

File: scripts/normalize_blog_asset_extensions.py
from __future__ import annotations

import re
import sys
from pathlib import Path


PATTERN = re.compile(
    r"(https://storage\.googleapis\.com/blogs_josa/sport/blog_assets/"
    r"[^\s\"')\]>}]+/img/thumbnail/[^\s\"')\]>}]+)\.(png|jpeg)"
)


def iter_qmd_files(paths: list[str]) -> list[Path]:
    files: list[Path] = []
    for raw_path in paths:
      path = Path(raw_path)
      if path.is_dir():
          files.extend(sorted(path.rglob("*.qmd")))
      elif path.is_file():
          files.append(path)
    return files


def main() -> int:
    roots = sys.argv[1:] or ["index.qmd", "posts"]

    for file_path in iter_qmd_files(roots):
        original = file_path.read_text(encoding="utf-8")
        updated = PATTERN.sub(r"\1.jpg", original)

        if updated != original:
            file_path.write_text(updated, encoding="utf-8")
            print(f"Updated {file_path}")

    return 0


if __name__ == "__main__":
    raise SystemExit(main())
